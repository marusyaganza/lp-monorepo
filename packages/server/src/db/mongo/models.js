const { formatIds, formatFilter } = require('../../utils/dbUtils');
const { ObjectId } = require('mongodb');

function createModel(db, table) {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result = await db.collection(table).find(formatFilter(filter));
    const data = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields) {
    const data = { ...fields };
    delete data.id;
    delete data.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    return result?.result;
  }

  async function deleteOne(filter) {
    console.log('filter', filter);
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    return result?.result;
  }

  async function createMany(entries) {
    const formattedEntries = entries.map(entry => ({
      ...entry,
      createdAt: Date.now()
    }));
    const result = await db.collection(table).insertMany(formattedEntries);
    return result.ops.map(formatIds);
  }

  return {
    findOne,
    findMany,
    createOne,
    createMany,
    updateOne,
    deleteOne
  };
}

module.exports = createModel;
