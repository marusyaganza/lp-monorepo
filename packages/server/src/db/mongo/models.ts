import { formatIds, formatFilter } from '../../utils/dbUtils';

import { Table, ModelType, DbData } from '..';

const { ObjectId } = require('mongodb');

function createModel<T extends DbData>(db: any, table: Table): ModelType<T> {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result: T[] = await db.collection(table).find(formatFilter(filter));
    const data: T[] = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields: Partial<T>) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields: Partial<T>) {
    const data = { ...fields };
    delete data?.id;
    delete data?.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    const isOk = result?.result?.ok ? true : false;
    const modifiedCount = result?.result?.nModified;
    return { ok: isOk, modifiedCount };
  }

  async function deleteOne(filter: Partial<T>) {
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    const isOk = result?.result?.ok ? true : false;
    return { ok: isOk };
  }

  return {
    findOne,
    findMany,
    createOne,
    updateOne,
    deleteOne
  };
}

export default createModel;
