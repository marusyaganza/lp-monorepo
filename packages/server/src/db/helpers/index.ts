export * from './sortingHelpers/sortingHelpers';
export * from './general/general';
