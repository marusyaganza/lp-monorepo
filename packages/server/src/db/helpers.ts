import { Document, Types } from 'mongoose';
import {
  Maybe,
  Language,
  Game,
  SortBy,
  SortWordsBy
} from '../generated/graphql';

type ModelDataType = {
  id?: Maybe<string>;
  user?: string;
};

interface DbData extends ModelDataType {
  _id?: string;
}

type DocumentType<T> =
  | (Document<unknown, unknown, T> &
      T & {
        _id: Types.ObjectId;
      })
  | null;

export interface WordsFilterType {
  sortBy?: SortBy | SortWordsBy | 'updatedAt';
  language: Language;
  isReverseOrder?: boolean;
  timesToLearn?: number | null;
  gameType?: Game;
  user?: string;
  tags?: string[];
  searchQuery?: string;
}

export interface WordsWithPaginationFilter extends WordsFilterType {
  page?: number;
  limit?: number;
}

export function formatData<T>(data: DocumentType<T>): T | null {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter?.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
const NO_TAGS_ID = '000000000000000000000000';

export function getWordsFilters(filter: WordsFilterType) {
  const {
    user,
    language = Language.English,
    sortBy,
    isReverseOrder,
    gameType,
    tags,
    searchQuery,
    timesToLearn = 5
  } = filter;

  // filter for a specific game
  let gameFilter;
  if (gameType === Game.Conjugation) {
    gameFilter = { conjugation: { $ne: null } };
  }
  if (gameType === Game.Audio) {
    gameFilter = { audioUrl: { $ne: '' } };
  }

  // filter based on tags
  let tagsFilters: { tags: string }[] = [];
  if (tags?.length) {
    tagsFilters = tags
      ?.filter(tag => tag !== NO_TAGS_ID)
      .map(tag => {
        return { tags: tag };
      });
  }

  let noTagsFilter;
  if (tags?.includes(NO_TAGS_ID)) {
    noTagsFilter = { $or: [{ tags: [] }, { tags: null }] };
  }

  let searchFilter;
  if (searchQuery) {
    const searchRegexp = { $regex: new RegExp(searchQuery, 'i') };
    searchFilter = {
      $or: [
        { name: searchRegexp },
        { 'defs.def': searchRegexp },
        { alternativeSpelling: searchRegexp },
        { stems: searchRegexp }
      ]
    };
  }

  // filter based on learning status
  let learningStatusFilters: Record<string, unknown>[] = [];
  if (gameType && sortBy !== SortBy.MemoryRefresher) {
    learningStatusFilters = [
      { isLearned: { $ne: true } },
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $lt: timesToLearn } },
          { [`statistics.${gameType}.successRate`]: null }
        ]
      }
    ];
  }
  if (sortBy === SortBy.MemoryRefresher) {
    learningStatusFilters = [
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $gte: timesToLearn } },
          { isLearned: true }
        ]
      }
    ];
  }

  //sorting
  let orderNum = isReverseOrder ? -1 : 1;

  // words with the highest number of errors should go first
  if (sortBy === SortBy.ErrorCount) {
    orderNum = -1 * orderNum;
  }

  let sort: Record<string, number> = { $natural: orderNum };

  if (sortBy) {
    let propName: string = sortBy;
    if (gameType) {
      propName = `statistics.${gameType}.${sortBy}`;
    }

    if (sortBy !== SortBy.MemoryRefresher) {
      sort = { [propName]: orderNum };
    }
  }

  const filtersArray = [
    gameFilter,
    noTagsFilter,
    searchFilter,
    ...tagsFilters,
    ...learningStatusFilters
  ].filter(Boolean);

  const filters: Record<string, unknown> = {
    user,
    language
  };

  if (filtersArray?.length) {
    filters.$and = filtersArray;
  }

  return {
    sort,
    filters
  };
}
