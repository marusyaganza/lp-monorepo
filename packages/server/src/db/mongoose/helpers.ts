import { Types, Document } from 'mongoose';
import { User, Word } from '../../generated/graphql';
import { DbData } from 'db';

type DocumentType =
  | (Document<unknown, Record<string, any>, User | Word> &
      Omit<
        | User
        | (Word & {
            _id: Types.ObjectId;
          }),
        never
      >)
  | null;

export function formatData(data: DocumentType) {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
