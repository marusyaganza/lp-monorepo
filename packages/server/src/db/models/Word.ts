import { QueryOptions } from 'mongoose';
import { Word } from '../schema/Word';
import { WordTag } from '../schema/WordTag';
import { Game } from '../schema/Game';

import {
  Word as WordType,
  NewWordInput,
  UpdateStatisticsInput,
  Language,
  Game as GameType,
  SortBy,
  SortWordsBy,
  WordStatisticsField
} from '../../generated/graphql';
import { formatFilter, formatData } from '../helpers';

const STATISTICS_FIELD: WordStatisticsField = {
  lastTimePracticed: 0,
  practicedTimes: 0,
  errorCount: 0,
  successRate: 0
};

const DEFAULT_STATISTICS: Record<GameType, WordStatisticsField> = {
  [GameType.Audio]: STATISTICS_FIELD,
  [GameType.SelectDef]: STATISTICS_FIELD,
  [GameType.SelectWord]: STATISTICS_FIELD,
  [GameType.TypeWord]: STATISTICS_FIELD,
  [GameType.Conjugation]: STATISTICS_FIELD
};

type wordsFilter = {
  sortBy?: SortBy | SortWordsBy | 'updatedAt';
  language: Language;
  isReverseOrder: boolean;
  timesToLearn?: number | null;
  gameType?: GameType;
  user?: string;
  tags?: string[];
};

export interface WordModelType {
  findOne: (filter: Partial<WordType>) => Promise<WordType | null>;
  findMany: (
    filter: Partial<WordType>,
    projection: string,
    options?: QueryOptions
  ) => Promise<WordType[] | null>;
  findManyAndSort: (filter: wordsFilter) => Promise<WordType[] | null>;
  createOne: (fields: NewWordInput) => Promise<WordType | null>;
  updateOne: (
    fields: Partial<WordType> & Pick<WordType, 'id' | 'user'>
  ) => Promise<{ ok: boolean; value?: WordType | null }>;
  updateMany: (
    fields: Partial<WordType> & Pick<WordType, 'id'>[],
    user?: string
  ) => Promise<{ ok: boolean }>;
  deleteOne: (filter: {
    id: string;
    user?: string;
  }) => Promise<{ ok: boolean }>;
  updateStatistics: (
    data: UpdateStatisticsInput[],
    user?: string
  ) => Promise<{ ok: boolean }>;
}

export const WordModel: WordModelType = {
  async findOne(filter) {
    if (!filter?.user) {
      return null;
    }
    const word = await Word.findOne(formatFilter(filter))
      .populate('tags')
      .exec();
    return formatData(word);
  },

  async findMany(filter, projection, options) {
    if (!filter?.user) {
      return [];
    }
    const words = await Word.find(formatFilter(filter), projection, options)
      .populate('tags')
      .exec();
    return words;
  },

  // Check if criteria is defined, if so => sort by criteria, else use natural sorting
  async findManyAndSort(filter) {
    const {
      user,
      language = Language.English,
      sortBy,
      isReverseOrder,
      gameType,
      tags,
      timesToLearn = 5
    } = filter;

    if (!user) {
      return [];
    }

    let orderNum = isReverseOrder ? -1 : 1;

    // words with the highest number of errors should go first
    if (sortBy === SortBy.ErrorCount) {
      orderNum = -1 * orderNum;
    }
    let learningStatusFilter = {};
    let sort: Record<string, number> = { $natural: orderNum };

    let gameFilter;

    if (gameType === GameType.Conjugation) {
      gameFilter = { conjugation: { $ne: null } };
    }

    if (gameType === GameType.Audio) {
      gameFilter = { audioUrl: { $ne: '' } };
    }

    let tagsFilters: { tags: string }[] = [];
    if (tags?.length) {
      tagsFilters = tags?.map(tag => {
        return { tags: tag };
      });
    }
    // TODO: refactor this part
    // Select words that are not learned or have been practiced without error less than 5 times in a row
    if (gameType && sortBy !== SortBy.MemoryRefresher) {
      learningStatusFilter = {
        $and: [
          { isLearned: { $ne: true } },
          ...tagsFilters,
          gameFilter,
          {
            $or: [
              { [`statistics.${gameType}.successRate`]: { $lt: timesToLearn } },
              { [`statistics.${gameType}.successRate`]: null }
            ]
          }
        ]
      };
    } else {
      if (tagsFilters?.length) {
        learningStatusFilter = {
          $and: tagsFilters
        };
      }
    }

    if (sortBy) {
      let propName: string = sortBy;
      if (gameType) {
        propName = `statistics.${gameType}.${sortBy}`;
      }

      if (sortBy === SortBy.MemoryRefresher) {
        learningStatusFilter = {
          $or: [
            { [`statistics.${gameType}.successRate`]: { $gte: timesToLearn } },
            { isLearned: true }
          ]
        };
      } else {
        sort = { [propName]: orderNum };
      }
    }

    const filters = {
      user,
      language,
      ...learningStatusFilter
    };

    const words = await Word.find(filters)
      // @ts-ignore
      .sort(sort)
      .populate('tags')
      .exec();

    return words;
  },

  async createOne(fields) {
    const createdAt = Date.now();
    const updatedAt = createdAt;
    const stems = fields?.stems?.length ? fields.stems : [fields.name];
    const transcription = fields?.transcription || fields?.name;
    const isOffensive = !!fields?.isOffensive;
    const word = await Word.create({
      ...fields,
      createdAt,
      updatedAt,
      stems,
      isOffensive,
      transcription,
      statistics: DEFAULT_STATISTICS
    });
    return formatData(word);
  },

  async updateOne(fields) {
    const updatedAt = Date.now();
    const word = await Word.findOne({ _id: fields.id, user: fields.user });
    if (!word) {
      return { ok: false };
    }
    const language = word.language;
    const { user, tags } = fields;
    let existingTags = await WordTag.find({ user, language }, 'id');
    existingTags = existingTags.map(tag => tag.id);
    // @ts-ignore
    const tagsUpdate = tags?.filter(tag => existingTags.includes(tag));
    const update = { ...fields, tags: tagsUpdate, updatedAt };
    const { acknowledged, modifiedCount } = await word.updateOne(update);
    const updatedWord = await Word.findById(word.id);
    const isOk = modifiedCount == 1 && acknowledged === true;
    return { ok: isOk, value: formatData(updatedWord) };
  },

  // This method do not have a practical use for now
  async updateMany(data, user) {
    let isOk = true;
    const updatedAt = Date.now();
    data.forEach(async entry => {
      const update = { ...entry, updatedAt };
      const result = await Word.findOneAndUpdate(
        { _id: entry.id, user },
        update,
        { includeResultMetadata: true, new: true }
      );
      isOk = isOk && result?.ok == 1 && result?.value !== null;
    });

    return { ok: isOk };
  },

  async updateStatistics(data) {
    let isOk = true;
    data.forEach(async entry => {
      try {
        const word = await Word.findById(entry.id);
        if (!word) {
          isOk = false;
          return;
        }
        if (!word?.statistics) {
          word.statistics = DEFAULT_STATISTICS;
        }

        const { gameType, hasError = false, isLearned = false } = entry;
        const gameConfig = await Game.findOne({ type: gameType });
        const timesToLearn = gameConfig?.timesToLearn || 5;

        const currentStatistics = word?.statistics?.[gameType];

        if (!currentStatistics) {
          word.statistics[gameType] = DEFAULT_STATISTICS[gameType];
        }

        const newStatistics = {
          practicedTimes: 1,
          errorCount: currentStatistics?.errorCount ?? 0,
          successRate: currentStatistics?.successRate ?? 0,
          lastTimePracticed: Date.now()
        };

        if (hasError) {
          newStatistics.errorCount++;
          newStatistics.successRate = 0;
        }

        if (!hasError) {
          newStatistics.successRate++;
        }

        if (isLearned) {
          newStatistics.successRate = timesToLearn;
        }

        if (currentStatistics?.practicedTimes) {
          newStatistics.practicedTimes += currentStatistics.practicedTimes;
        }

        word.statistics[gameType] = newStatistics;
        await word.save();
      } catch (err) {
        console.error('saving statisticks failed', err);
        isOk = false;
      }
    });
    return { ok: isOk };
  },

  async deleteOne(filter) {
    const result = { ok: false };
    const { deletedCount, acknowledged } = await Word.deleteOne({
      _id: filter.id,
      user: filter.user
    });
    if (acknowledged && deletedCount == 1) {
      result.ok = true;
    }
    return result;
  }
};
