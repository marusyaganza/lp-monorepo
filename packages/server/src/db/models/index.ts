import { UserModel, UserModelType } from './User';
import { WordModel, WordModelType } from './Word';
import { WordTagModel, WordTagModelType } from './WordTag';

export interface ModelsType {
  User: UserModelType;
  Word: WordModelType;
  WordTag: WordTagModelType;
}

export const models: ModelsType = {
  User: UserModel,
  Word: WordModel,
  WordTag: WordTagModel
};
