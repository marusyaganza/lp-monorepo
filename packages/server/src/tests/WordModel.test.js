import { WordModel } from '../db/models/Word';
import { connectToDb, disconnectFromDb, dropDb } from './helpers';
import { testData } from './mocks/dbTestData';
import { Game, SortBy, SortWordsBy } from '../generated/graphql';

const snapshotConfig = {
  createdAt: expect.any(String),
  updatedAt: expect.any(Number),
  id: expect.any(String),
  _id: expect.any(Object)
};

const mockUser = { id: '6480560e8cad1841ed6b4011', role: 'MEMBER' };
const mockId = '6480653341e6f90377d19cfb';

let index = 0;

jest.spyOn(global.Date, 'now').mockImplementation(function () {
  return index++;
});

describe('WordModel', () => {
  beforeAll(async () => {
    await connectToDb();
    await dropDb();
  });

  afterEach(async () => {
    await dropDb();
  });

  afterAll(async () => {
    await disconnectFromDb();
  });

  test('createOne', async () => {
    const result = await WordModel.createOne({
      ...testData.createWordInput,
      user: mockUser.id
    });
    expect(result).toMatchSnapshot(snapshotConfig);
  });

  test('findOne', async () => {
    const word = await WordModel.createOne({
      ...testData.createWordInput,
      user: mockUser.id
    });
    await WordModel.createOne({
      ...testData.createWordInput,
      user: mockId
    });
    const result = await WordModel.findOne({
      id: word.id,
      user: mockUser.id
    });
    const result2 = await WordModel.findOne({
      uuid: testData.createWordInput.uuid,
      user: mockUser.id
    });
    expect(result2).not.toBeNull();
    expect(result.id).toEqual(result2.id);
    expect(result.user).toEqual(result2.user);
    expect(result).toMatchSnapshot(snapshotConfig);
  });

  test('findOne should return null if user id is not provided', async () => {
    const word = await WordModel.createOne({
      ...testData.createWordInput,
      user: mockUser.id
    });
    const result = await WordModel.findOne({
      id: word.id
    });
    const result2 = await WordModel.findOne({
      uuid: testData.createWordInput.uuid
    });
    expect(result).toBeNull();
    expect(result2).toBeNull();
  });

  test('findMany', async () => {
    await WordModel.createOne({
      ...testData.createWordInput,
      user: mockUser.id
    });
    await WordModel.createOne({
      ...testData.createWordInput2,
      user: mockUser.id
    });
    await WordModel.createOne({
      ...testData.createWordInput,
      user: mockId
    });

    const resultForUnknowUser = await WordModel.findMany({});
    const result = await WordModel.findMany({ user: mockUser.id });
    expect(result).toHaveLength(2);
    expect(resultForUnknowUser).toHaveLength(0);
  });

  test('findManyAndSort with reverse order', async () => {
    await WordModel.createOne({
      ...testData.createWordInput,
      user: mockUser.id
    });
    await WordModel.createOne({
      ...testData.createWordInput2,
      user: mockUser.id
    });
    await WordModel.createOne({
      ...testData.createWordInput,
      user: mockId
    });

    const resultForUnknowUser = await WordModel.findManyAndSort({});
    const result = await WordModel.findManyAndSort({
      user: mockUser.id,
      isReverseOrder: true
    });
    expect(result).toHaveLength(2);
    const index = result.findIndex(
      item => item.uuid === testData.createWordInput.uuid
    );
    expect(index).toEqual(1);
    expect(resultForUnknowUser).toHaveLength(0);
  });

  test('findManyAndSort by name, level, and date', async () => {
    await WordModel.createOne({
      ...testData.createWordInput,
      user: mockUser.id
    });
    await WordModel.createOne({
      ...testData.createWordInput2,
      user: mockUser.id
    });
    await WordModel.createOne({
      ...testData.createWordInput,
      user: mockId
    });

    const resultByName = await WordModel.findManyAndSort({
      user: mockUser.id,
      sortBy: SortWordsBy.Name
    });
    expect(resultByName).toHaveLength(2);
    const indexByName = resultByName.findIndex(
      item => item.uuid === testData.createWordInput.uuid
    );
    expect(indexByName).toEqual(0);

    const resultByLevel = await WordModel.findManyAndSort({
      user: mockUser.id,
      sortBy: SortWordsBy.Level
    });
    expect(resultByLevel).toHaveLength(2);
    const indexByLevel = resultByName.findIndex(
      item => item.uuid === testData.createWordInput.uuid
    );
    expect(indexByLevel).toEqual(0);

    const resultByDate = await WordModel.findManyAndSort({
      user: mockUser.id,
      sortBy: 'updatedAt'
    });
    expect(resultByDate).toHaveLength(2);
    const indexByDate = resultByName.findIndex(
      item => item.uuid === testData.createWordInput.uuid
    );
    expect(indexByDate).toEqual(0);
  });

  test('findManyAndSort based on game statistics', async () => {
    await WordModel.createOne({
      ...testData.createWordInput,
      user: mockUser.id
    });
    const createdWord = await WordModel.createOne({
      ...testData.createWordInput2,
      user: mockUser.id
    });

    const updateResult = await WordModel.updateStatistics(
      [{ id: createdWord.id, hasError: true, gameType: Game.Audio }],
      mockUser.id
    );

    const result = await WordModel.findManyAndSort({
      user: mockUser.id,
      sortBy: SortBy.ErrorCount
    });
    expect(result).toHaveLength(2);
    expect(updateResult).toEqual({ ok: true });
    const index = result.findIndex(item => item.id === createdWord.id);
    expect(index).toEqual(1);
  });

  test('UpdateStatistics 2 times', async () => {
    await WordModel.createOne({
      ...testData.createWordInput,
      user: mockUser.id
    });
    const createdWord = await WordModel.createOne({
      ...testData.createWordInput2,
      user: mockUser.id
    });

    await WordModel.updateStatistics(
      [{ id: createdWord.id, hasError: true, gameType: Game.Audio }],
      mockUser.id
    );

    await WordModel.updateStatistics(
      [{ id: createdWord.id, hasError: false, gameType: Game.Audio }],
      mockUser.id
    );

    const result = await WordModel.findManyAndSort({
      user: mockUser.id,
      sortBy: SortBy.PracticedTimes
    });

    const index = result.findIndex(item => item.id === createdWord.id);
    const regularWord = result[0];

    expect(index).toEqual(1);
    expect(regularWord.statistics.AUDIO.practicedTimes).toEqual(0);
    expect(regularWord.statistics.AUDIO.errorCount).toEqual(0);
  });

  test('updateOne', async () => {
    const word = await WordModel.createOne({
      ...testData.createWordInput,
      user: mockUser.id
    });
    const word2 = await WordModel.createOne({
      ...testData.createWordInput2,
      user: mockUser.id
    });

    const result = await WordModel.updateOne({
      ...testData.updateWordInput,
      user: mockUser.id,
      id: word.id
    });

    expect(result.ok).toBe(true);
    expect(result.value).toMatchSnapshot(snapshotConfig);

    const searchTResult = await WordModel.findOne({
      id: word.id,
      user: mockUser.id
    });

    expect(result.value).toEqual(searchTResult);

    const resultForUnknowUser = await WordModel.updateOne({
      ...testData.updateWordInput,
      id: word2.id
    });
    expect(resultForUnknowUser).toEqual({ ok: false, value: null });

    const resultWithImmutableProps = await WordModel.updateOne({
      ...testData.updateWordInput2,
      user: mockUser.id,
      id: word2.id
    });

    expect(resultWithImmutableProps.value.name).toEqual(word2.name);
    expect(resultWithImmutableProps.value.uuid).toEqual(word2.uuid);
  });

  test('deleteOne', async () => {
    const word = await WordModel.createOne({
      ...testData.createWordInput,
      user: mockUser.id
    });
    await WordModel.createOne({
      ...testData.createWordInput2,
      user: mockUser.id
    });
    await WordModel.createOne({
      ...testData.createWordInput,
      user: mockUser.id
    });
    const result = await WordModel.deleteOne({
      id: word.id,
      user: mockUser.id
    });
    expect(result).toEqual({ ok: true });
    const searchTResult = await WordModel.findOne({
      id: word.id,
      user: mockUser.id
    });
    const words = await WordModel.findMany({ user: mockUser.id });
    expect(words).toHaveLength(2);
    expect(searchTResult).toBeNull();

    const resultForUnknowUser = await WordModel.deleteOne({
      id: word.id
    });
    expect(resultForUnknowUser).toEqual({ ok: false });
  });
});
