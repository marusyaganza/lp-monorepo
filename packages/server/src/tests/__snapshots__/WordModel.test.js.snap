// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordModel createOne 1`] = `
{
  "_id": Any<Object>,
  "additionalInfo": "additional info",
  "audioUrl": "mock.mp3",
  "createdAt": Any<String>,
  "defs": [
    {
      "def": "a person who believes that everyone is equal and should have the same rights and opportunities",
      "examples": [
        {
          "text": "He described himself as ‘an {it}egalitarian{/it}’.",
        },
      ],
    },
    {
      "def": "aiming for equal wealth, status, etc., for all people",
      "examples": [
        {
          "text": "He is a committed {it}egalitarian{/it}",
        },
        {
          "text": "Even in the {it}egalitarian{/it} Nordic country, Marin felt hergender and age sometimes received too much emphasis. Kostya Manenkov And Karl Ritter, USA TODAY, 21 Aug. 2022",
        },
      ],
    },
  ],
  "id": Any<String>,
  "imgUrl": "mock.jpeg",
  "isOffensive": false,
  "language": "ENGLISH",
  "level": "C1",
  "name": "egalitarian",
  "particle": "noun",
  "shortDef": [
    "aiming for equal wealth, status, etc., for all people",
  ],
  "stems": [
    "egalitarian",
    "egalitarianism",
    "egalitarians",
  ],
  "tags": [
    {
      "color": "pink",
      "text": "academic",
    },
  ],
  "transcription": "iˌɡælɪˈteəriən",
  "user": "6480560e8cad1841ed6b4011",
  "uuid": "74046e79-e4c9-4b52-ac96-cb7ae98fb601",
}
`;

exports[`WordModel findOne 1`] = `
{
  "_id": Any<Object>,
  "additionalInfo": "additional info",
  "audioUrl": "mock.mp3",
  "createdAt": Any<String>,
  "defs": [
    {
      "def": "a person who believes that everyone is equal and should have the same rights and opportunities",
      "examples": [
        {
          "text": "He described himself as ‘an {it}egalitarian{/it}’.",
        },
      ],
    },
    {
      "def": "aiming for equal wealth, status, etc., for all people",
      "examples": [
        {
          "text": "He is a committed {it}egalitarian{/it}",
        },
        {
          "text": "Even in the {it}egalitarian{/it} Nordic country, Marin felt hergender and age sometimes received too much emphasis. Kostya Manenkov And Karl Ritter, USA TODAY, 21 Aug. 2022",
        },
      ],
    },
  ],
  "id": Any<String>,
  "imgUrl": "mock.jpeg",
  "isOffensive": false,
  "language": "ENGLISH",
  "level": "C1",
  "name": "egalitarian",
  "particle": "noun",
  "shortDef": [
    "aiming for equal wealth, status, etc., for all people",
  ],
  "stems": [
    "egalitarian",
    "egalitarianism",
    "egalitarians",
  ],
  "tags": [
    {
      "color": "pink",
      "text": "academic",
    },
  ],
  "transcription": "iˌɡælɪˈteəriən",
  "user": "6480560e8cad1841ed6b4011",
  "uuid": "74046e79-e4c9-4b52-ac96-cb7ae98fb601",
}
`;

exports[`WordModel updateOne 1`] = `
{
  "_id": Any<Object>,
  "additionalInfo": "very useful information",
  "audioUrl": "mock.mp3",
  "createdAt": Any<String>,
  "defs": [
    {
      "def": "{bc} a strong substance that stretches and that is made out of chemicals or from the juice of a tropical tree",
      "examples": [
        {
          "text": "a rubber plantation",
        },
        {
          "text": "a rubber tree",
        },
      ],
    },
    {
      "def": "{b}{it}rubbers{/it}{/b} {it}US{/it}, {it}informal + old-fashioned{/it} {bc} shoes or boots that are made of rubber and that fit over your regular shoes to keep them dry",
      "examples": [],
    },
  ],
  "id": Any<String>,
  "imgUrl": "mock.jpeg",
  "isOffensive": false,
  "language": "ENGLISH",
  "level": "B2",
  "name": "egalitarian",
  "particle": "noun",
  "shortDef": [
    "aiming for equal wealth, status, etc., for all people",
  ],
  "stems": [
    "egalitarian",
    "egalitarianism",
    "egalitarians",
  ],
  "tags": [
    {
      "color": "pink",
      "text": "academic",
    },
  ],
  "transcription": "iˌɡælɪˈteəriən",
  "user": "6480560e8cad1841ed6b4011",
  "uuid": "74046e79-e4c9-4b52-ac96-cb7ae98fb601",
}
`;
